File: idp/src/infra/base64.ts
// Copyright (c) 2023 Anass Bouassaba.
//
// Use of this software is governed by the Business Source License
// included in the file LICENSE in the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the GNU Affero General Public License v3.0 only, included in the file
// AGPL-3.0-only in the root of this repository.
import { Buffer } from 'node:buffer'

export function base64ToBuffer(value: string): Buffer | null {
  let withoutPrefix: string
  if (value.includes(',')) {
    withoutPrefix = value.split(',')[1]
  } else {
    withoutPrefix = value
  }
  try {
    return Buffer.from(withoutPrefix, 'base64')
  } catch {
    return null
  }
}

export function base64ToMIME(value: string): string | null {
  if (!value.startsWith('data:image/')) {
    return ''
  }
  const colonIndex = value.indexOf(':')
  const semicolonIndex = value.indexOf(';')
  if (colonIndex === -1 || semicolonIndex === -1) {
    return ''
  }
  return value.substring(colonIndex + 1, semicolonIndex)
}

export function base64ToExtension(value: string): string {
  const mime = base64ToMIME(value)
  switch (mime) {
    case 'image/jpeg':
      return '.jpg'
    case 'image/png':
      return '.png'
    default:
      return ''
  }
}

File: shared/helper/base64.go
// Copyright (c) 2023 Anass Bouassaba.
//
// Use of this software is governed by the Business Source License
// included in the file LICENSE in the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the GNU Affero General Public License v3.0 only, included in the file
// AGPL-3.0-only in the root of this repository.

package helper

import (
	"encoding/base64"
	"fmt"
	"mime"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

func Base64ToBytes(value string) ([]byte, error) {
	var withoutPrefix string
	if strings.Contains(value, ",") {
		withoutPrefix = strings.Split(value, ",")[1]
	} else {
		withoutPrefix = value
	}
	b, err := base64.StdEncoding.DecodeString(withoutPrefix)
	if err != nil {
		return nil, err
	}
	return b, nil
}

func Base64ToMIME(value string) string {
	if !strings.HasPrefix(value, "data:image/") {
		return ""
	}
	colonIndex := strings.Index(value, ":")
	semicolonIndex := strings.Index(value, ";")
	if colonIndex == -1 || semicolonIndex == -1 {
		return ""
	}
	return value[colonIndex+1 : semicolonIndex]
}

func Base64ToExtension(value string) string {
	switch Base64ToMIME(value) {
	case "image/jpeg":
		return ".jpg"
	case "image/png":
		return ".png"
	default:
		return ""
	}
}

func FileToBase64(path string) (*string, error) {
	data, err := os.ReadFile(path) //nolint:gosec // Safe path
	if err != nil {
		return nil, err
	}
	mimeType := mime.TypeByExtension(filepath.Ext(path))
	if mimeType == "" {
		mimeType = http.DetectContentType(data)
	}
	base64 := fmt.Sprintf("data:%s;base64,%s", mimeType, base64.StdEncoding.EncodeToString(data))
	return &base64, nil
}
